File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] arr) {
           
        int[] nge = new int[arr.length];
        int[] pge = new int[arr.length];
         Stack<Integer> st = new Stack<>();
   
        for(int i=0;i<arr.length;i++){
            while(st.size() > 0 && arr[st.peek()] > arr[i])st.pop();
            if(st.size() == 0) pge[i] = -1;
            else pge[i] = st.peek();
            st.add(i);
            
        }

        while(st.size() > 0) st.pop();

         for(int i=arr.length-1;i>=0;i--){
            while(st.size() > 0 && arr[st.peek()] >= arr[i])st.pop();
            if(st.size() == 0) nge[i] = arr.length;
            else nge[i] = st.peek();
            st.add(i);
            
        }
         int ans = 0;
        for(int i=0;i<arr.length;i++){
           ans =  Math.max(ans,((nge[i] - pge[i] -1) * arr[i]));
        }
        return ans;
       

    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    void findNSE(vector<int>& nse, vector<int>& heights) {
        stack<int> st;
        for (int i = heights.size() - 1; i >= 0; i--) {
            while (!st.empty() && heights[st.top()] >= heights[i]) {
                st.pop();
            }
            if (!st.empty()) {
                nse[i] = st.top();
            } else {
                nse[i] = heights.size(); // No smaller element, set to end of array
            }
            st.push(i);
        }
    }

    void findPGE(vector<int>& pge, vector<int>& heights) {
        stack<int> st;
        for (int i = 0; i < heights.size(); i++) {
            while (!st.empty() && heights[st.top()] >= heights[i]) {
                st.pop();
            }
            if (!st.empty()) {
                pge[i] = st.top();
            } else {
                pge[i] = -1; // No greater element, set to start of array
            }
            st.push(i);
        }
    }

    int largestRectangleArea(vector<int>& heights) {
        vector<int> nse(heights.size(), heights.size());
        vector<int> pge(heights.size(), -1);
        
        findNSE(nse, heights);
        findPGE(pge, heights);

        int maxi = 0;
        for (int i = 0; i < heights.size(); i++) {
            int base = nse[i] - pge[i] - 1; // Correct base calculation
            int area = heights[i] * base;
            maxi = max(maxi, area);
        }
        return maxi;
    }
};
